## **BÀI TẬP CHƯƠNG 2**

### **1. Tạo hai bẳng dữ liệu**

In [3]:
import sqlite3

# Kết nối đến SQLite (tạo file nếu chưa có)
conn = sqlite3.connect("university.db")
cursor = conn.cursor()

# Tạo bảng course
cursor.execute('''
    CREATE TABLE IF NOT EXISTS course (
        id INTEGER PRIMARY KEY,
        course_name TEXT NOT NULL
    )
''')

# Tạo bảng student
cursor.execute('''
    CREATE TABLE IF NOT EXISTS student (
        student_id INTEGER PRIMARY KEY,
        name TEXT NOT NULL,
        class TEXT NOT NULL,
        course_id INTEGER,
        score REAL,
        FOREIGN KEY (course_id) REFERENCES course(id)
    )
''')

# Lưu thay đổi
conn.commit()
conn.close()


In [ ]:
conn = sqlite3.connect("university.db")
cursor = conn.cursor()

# Chèn dữ liệu vào bảng course
courses = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]
cursor.executemany("INSERT INTO course (id, course_name) VALUES (?, ?)", courses)

# Chèn dữ liệu vào bảng student
students = [
    (1, "Nguyen Minh Hoang", "May Tinh", 12, 6.7),
    (2, "Tran Thi Lan", "Kinh Te", 34, 9.2),
    (3, "Pham Van Nam", "Toan Tin", None, 7.9),
    (4, "Le Thanh Huyen", "Toan Tin", 20, 7.2),
    (5, "Vu Quoc Anh", "May Tinh", 24, 8.0),
    (6, "Dang Thuy Linh", "May Tinh", 24, 5.5),
    (7, "Bui Tien Dung", "Kinh Te", 34, 9.2),
    (8, "Ho Ngoc Mai", "Toan Tin", 20, 8.8),
    (9, "Duong Huu Phuc", "Kinh Te", None, 7.2),
    (10, "Cao Thi Hanh", "May Tinh", None, 7.0)
]
cursor.executemany("INSERT INTO student (student_id, name, class, course_id, score) VALUES (?, ?, ?, ?, ?)", students)

conn.commit()
conn.close()


In [4]:
conn = sqlite3.connect("university.db")
cursor = conn.cursor()
import pandas as pd
# Hiển thị dữ liệu ban đầu
def show_table(query):
    df = pd.read_sql_query(query, conn)
    display(df)

print("Dữ liệu bảng student:")
show_table("SELECT * FROM student")

print("Dữ liệu bảng course:")
show_table("SELECT * FROM course")

Dữ liệu bảng student:


student_id               name     class  course_id  score
0           1  Nguyen Minh Hoang  May Tinh       12.0    6.7
1           2       Tran Thi Lan   Kinh Te       34.0    9.2
2           3       Pham Van Nam  Toan Tin        NaN    7.9
3           4     Le Thanh Huyen  Toan Tin       20.0    7.2
4           5        Vu Quoc Anh  May Tinh       24.0    8.0
5           6     Dang Thuy Linh  May Tinh       24.0    5.5
6           7      Bui Tien Dung   Kinh Te       34.0    9.2
7           8        Ho Ngoc Mai  Toan Tin       20.0    8.8
8           9     Duong Huu Phuc   Kinh Te        NaN    7.2
9          10       Cao Thi Hanh  May Tinh        NaN    7.0

Dữ liệu bảng course:


id course_name
0  12   Giai tich
1  26     Tin hoc
2  34    Thong ke

#### **2. Hãy kết nối hai bảng trên theo các cách**

**2.1.  Sử dụng tích Decartes**

In [5]:
print("Tích Descartes:")
df_descartes = pd.read_sql_query("SELECT * FROM student CROSS JOIN course", conn)
display(df_descartes)


Tích Descartes:


student_id               name     class  course_id  score  id course_name
0            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  12   Giai tich
1            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  26     Tin hoc
2            1  Nguyen Minh Hoang  May Tinh       12.0    6.7  34    Thong ke
3            2       Tran Thi Lan   Kinh Te       34.0    9.2  12   Giai tich
4            2       Tran Thi Lan   Kinh Te       34.0    9.2  26     Tin hoc
5            2       Tran Thi Lan   Kinh Te       34.0    9.2  34    Thong ke
6            3       Pham Van Nam  Toan Tin        NaN    7.9  12   Giai tich
7            3       Pham Van Nam  Toan Tin        NaN    7.9  26     Tin hoc
8            3       Pham Van Nam  Toan Tin        NaN    7.9  34    Thong ke
9            4     Le Thanh Huyen  Toan Tin       20.0    7.2  12   Giai tich
10           4     Le Thanh Huyen  Toan Tin       20.0    7.2  26     Tin hoc
11           4     Le Thanh Huyen  Toan Tin       20.0    7.2  34    Thong ke
12           5        Vu Quoc Anh  May Tinh       24.0    8.0  12   Giai tich
13           5        Vu Quoc Anh  May Tinh       24.0    8.0  26     Tin hoc
14           5        Vu Quoc Anh  May Tinh       24.0    8.0  34    Thong ke
15           6     Dang Thuy Linh  May Tinh       24.0    5.5  12   Giai tich
16           6     Dang Thuy Linh  May Tinh       24.0    5.5  26     Tin hoc
17           6     Dang Thuy Linh  May Tinh       24.0    5.5  34    Thong ke
18           7      Bui Tien Dung   Kinh Te       34.0    9.2  12   Giai tich
19           7      Bui Tien Dung   Kinh Te       34.0    9.2  26     Tin hoc
20           7      Bui Tien Dung   Kinh Te       34.0    9.2  34    Thong ke
21           8        Ho Ngoc Mai  Toan Tin       20.0    8.8  12   Giai tich
22           8        Ho Ngoc Mai  Toan Tin       20.0    8.8  26     Tin hoc
23           8        Ho Ngoc Mai  Toan Tin       20.0    8.8  34    Thong ke
24           9     Duong Huu Phuc   Kinh Te        NaN    7.2  12   Giai tich
25           9     Duong Huu Phuc   Kinh Te        NaN    7.2  26     Tin hoc
26           9     Duong Huu Phuc   Kinh Te        NaN    7.2  34    Thong ke
27          10       Cao Thi Hanh  May Tinh        NaN    7.0  12   Giai tich
28          10       Cao Thi Hanh  May Tinh        NaN    7.0  26     Tin hoc
29          10       Cao Thi Hanh  May Tinh        NaN    7.0  34    Thong ke

- **Tích Descartes (Cartesian Product):**
  - Kết hợp tất cả các hàng của bảng `student` với tất cả các hàng của bảng `course`.
  - **Kết quả:** 30 hàng (10 sinh viên × 3 khóa học).

**2.2. Sử dụng JOIN: INNER JOIN, LEFT JOIN, RIGHT JOIN, FULL OUTER JOIN**

In [6]:
# INNER JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
INNER JOIN course c ON s.course_id = c.id
''')
inner_join = cursor.fetchall()
print("INNER JOIN:")
for row in inner_join:
    print(row)

INNER JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)


  - **INNER JOIN:** Chỉ lấy các hàng có khớp `course_id` trong cả hai bảng. **Kết quả:** 3 hàng.

In [7]:
# LEFT JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
LEFT JOIN course c ON s.course_id = c.id
''')
left_join = cursor.fetchall()
print("LEFT JOIN:")
for row in left_join:
    print(row)

LEFT JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9)
(4, 'Le Thanh Huyen', 'Toan Tin', None, 7.2)
(5, 'Vu Quoc Anh', 'May Tinh', None, 8.0)
(6, 'Dang Thuy Linh', 'May Tinh', None, 5.5)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(8, 'Ho Ngoc Mai', 'Toan Tin', None, 8.8)
(9, 'Duong Huu Phuc', 'Kinh Te', None, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', None, 7.0)


 - **LEFT JOIN:** Lấy tất cả các hàng từ bảng `student` và khớp với bảng `course` nếu có. **Kết quả:** 10 hàng.

In [8]:
# RIGHT JOIN
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
RIGHT JOIN course c ON s.course_id = c.id
''')
right_join = cursor.fetchall()
print("RIGHT JOIN:")
for row in right_join:
    print(row)

RIGHT JOIN:
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(None, None, None, 'Tin hoc', None)


  - **RIGHT JOIN:** Lấy tất cả các hàng từ bảng `course` và khớp với bảng `student` nếu có. **Kết quả:** 4 hàng.

In [9]:
# FULL OUTER JOIN (SQLite không hỗ trợ FULL OUTER JOIN, sử dụng UNION của LEFT JOIN và RIGHT JOIN)
cursor.execute('''
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
LEFT JOIN course c ON s.course_id = c.id
UNION
SELECT s.student_id, s.name, s.class, c.course_name, s.score
FROM student s
RIGHT JOIN course c ON s.course_id = c.id
''')
full_outer_join = cursor.fetchall()
print("FULL OUTER JOIN:")
for row in full_outer_join:
    print(row)

FULL OUTER JOIN:
(None, None, None, 'Tin hoc', None)
(1, 'Nguyen Minh Hoang', 'May Tinh', 'Giai tich', 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 'Thong ke', 9.2)
(3, 'Pham Van Nam', 'Toan Tin', None, 7.9)
(4, 'Le Thanh Huyen', 'Toan Tin', None, 7.2)
(5, 'Vu Quoc Anh', 'May Tinh', None, 8.0)
(6, 'Dang Thuy Linh', 'May Tinh', None, 5.5)
(7, 'Bui Tien Dung', 'Kinh Te', 'Thong ke', 9.2)
(8, 'Ho Ngoc Mai', 'Toan Tin', None, 8.8)
(9, 'Duong Huu Phuc', 'Kinh Te', None, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', None, 7.0)


  - **FULL OUTER JOIN:** Lấy tất cả các hàng từ cả hai bảng, bao gồm cả các hàng không khớp. **Kết quả:** 11 hàng.

### **3.Hãy cập nhật những giá trị course_id còn thiếu trong bảng student bằng câu lệnh SQL,trong đó các giá trị được điền là những giá trị nằm trong bảng course và loại bỏ những bản ghi tham gia những môn học không tồn tại bảng course**

In [10]:
# Điền giá trị course_id còn thiếu bằng cách chọn ngẫu nhiên từ bảng course
cursor.execute('''
UPDATE student
SET course_id = (SELECT id FROM course ORDER BY RANDOM() LIMIT 1)
WHERE course_id IS NULL
''')
conn.commit()

In [11]:
# Loại bỏ các bản ghi có course_id không tồn tại trong bảng course
cursor.execute('''
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course)
''')
conn.commit()

In [14]:
# Hiển thị kết quả sau khi cập nhật
cursor.execute('SELECT * FROM student')
updated_students = cursor.fetchall()

print("Danh sách sinh viên sau khi cập nhật:")
for row in updated_students:
    print(row)

Danh sách sinh viên sau khi cập nhật:
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7)
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2)
(3, 'Pham Van Nam', 'Toan Tin', 26, 7.9)
(7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2)
(9, 'Duong Huu Phuc', 'Kinh Te', 26, 7.2)
(10, 'Cao Thi Hanh', 'May Tinh', 26, 7.0)


- **Cập nhật giá trị `course_id` còn thiếu:**
  - Sử dụng giá trị ngẫu nhiên từ bảng `course` để điền vào các giá trị `NULL`.
  - *Ví dụ* Cao Thi Hanh (ban đầu `course_id = NULL`) được gán `course_id = 26`.

- **Loại bỏ các bản ghi không hợp lệ:**
  - Xóa các hàng có `course_id` không tồn tại trong bảng `course`.

**3.1. Tổng số sinh viên, điểm trung bình của từng lớp.**

In [28]:
cursor.execute('''
SELECT class, COUNT(student_id) AS total_students, AVG(score) AS average_score
FROM student
GROUP BY class
''')
class_stats = cursor.fetchall()
print("Thống kê theo lớp:")
for row in class_stats:
    print(row)

Thống kê theo lớp:
('Kinh Te', 3, 8.533333333333333)
('May Tinh', 2, 6.85)
('Toan Tin', 1, 7.9)


- Lớp Kinh Tế có nhiều sinh viên nhất (3 sinh viên) và điểm trung bình cao nhất (8.53).
- Lớp Máy Tính có điểm trung bình thấp nhất (6.85).
- Lớp Toán Tin chỉ có 1 sinh viên nên điểm trung bình chính là điểm của sinh viên đó (7.9).

**3.2. Tổng số sinh viên, điểm trung bình của từng môn học**

In [29]:
cursor.execute('''
SELECT c.course_name, COUNT(s.student_id) AS total_students, AVG(s.score) AS average_score
FROM student s
JOIN course c ON s.course_id = c.id
GROUP BY c.course_name
''')
course_stats = cursor.fetchall()
print("Thống kê theo môn học:")
for row in course_stats:
    print(row)

Thống kê theo môn học:
('Giai tich', 1, 6.7)
('Thong ke', 2, 9.2)
('Tin hoc', 3, 7.366666666666667)


- Môn Thống kê có điểm trung bình cao nhất (9.2).
- Môn Giải tích có điểm trung bình thấp nhất (6.7) nhưng chỉ có 1 sinh viên, nên chưa phản ánh hết thực tế.
- Môn Tin học có nhiều sinh viên nhất (3 người) và điểm trung bình ở mức trung bình (7.37).

### 3.3. Phân loại thi đua theo số điểm của từng môn học

Biết:

- **Xuất sắc**: Điểm trung bình **≥ 9.0**  
- **Tốt**: Điểm trung bình **từ 6.0 đến 8.9**  
- **Kém**: Điểm trung bình **< 6.0**  


In [16]:
cursor.execute('''
SELECT c.course_name, AVG(s.score) AS average_score,
       CASE
           WHEN AVG(s.score) >= 9.0 THEN 'Xuất sắc'
           WHEN AVG(s.score) >= 6.0 AND AVG(s.score) < 9.0 THEN 'Tốt'
           ELSE 'Kém'
       END AS performance
FROM student s
JOIN course c ON s.course_id = c.id
GROUP BY c.course_name
''')
performance_stats = cursor.fetchall()
print("Phân loại thi đua:")
for row in performance_stats:
    print(row)

Phân loại thi đua:
('Giai tich', 6.7, 'Tốt')
('Thong ke', 9.2, 'Xuất sắc')
('Tin hoc', 7.366666666666667, 'Tốt')


### **4.Hãy xếp hạng sinh viên thông qua:**

    a. Điểm số.
    b. Điểm số theo lớp học.
    c. Điểm số theo mã môn học.
và cho biết top 3 sinh viện đạt thứ hạng cao nhất, top 3 sinh viên đạt thứ hạng thấp nhất theo
từng trường hợp trên.


In [17]:
#xếp hạng theo điểm số
cursor.execute('''
SELECT student_id, name, score,
       RANK() OVER (ORDER BY score DESC) AS rank_overall
FROM student
''')
rank_overall = cursor.fetchall()
print("Xếp hạng theo điểm số:")
for row in rank_overall:
    print(row)

Xếp hạng theo điểm số:
(2, 'Tran Thi Lan', 9.2, 1)
(7, 'Bui Tien Dung', 9.2, 1)
(3, 'Pham Van Nam', 7.9, 3)
(9, 'Duong Huu Phuc', 7.2, 4)
(10, 'Cao Thi Hanh', 7.0, 5)
(1, 'Nguyen Minh Hoang', 6.7, 6)


- Số 2 và số 7 có điểm số cao nhất (9.2)
- Số 1 có điểm số thấp nhất (6.7)

In [18]:
#xếp hạng theo số lớp học
cursor.execute('''
SELECT student_id, name, class, score,
       RANK() OVER (PARTITION BY class ORDER BY score DESC) AS rank_by_class
FROM student
''')
rank_by_class = cursor.fetchall()
print("Xếp hạng theo lớp học:")
for row in rank_by_class:
    print(row)

Xếp hạng theo lớp học:
(2, 'Tran Thi Lan', 'Kinh Te', 9.2, 1)
(7, 'Bui Tien Dung', 'Kinh Te', 9.2, 1)
(9, 'Duong Huu Phuc', 'Kinh Te', 7.2, 3)
(10, 'Cao Thi Hanh', 'May Tinh', 7.0, 1)
(1, 'Nguyen Minh Hoang', 'May Tinh', 6.7, 2)
(3, 'Pham Van Nam', 'Toan Tin', 7.9, 1)


In [19]:
cursor.execute('''
SELECT student_id, name, course_id, score,
       RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS rank_by_course
FROM student
''')
rank_by_course = cursor.fetchall()
print("Xếp hạng theo mã môn học:")
for row in rank_by_course:
    print(row)

Xếp hạng theo mã môn học:
(1, 'Nguyen Minh Hoang', 12, 6.7, 1)
(3, 'Pham Van Nam', 26, 7.9, 1)
(9, 'Duong Huu Phuc', 26, 7.2, 2)
(10, 'Cao Thi Hanh', 26, 7.0, 3)
(2, 'Tran Thi Lan', 34, 9.2, 1)
(7, 'Bui Tien Dung', 34, 9.2, 1)


In [20]:
# Top 3 cao nhất
cursor.execute('''
SELECT student_id, name, score
FROM student
ORDER BY score DESC
LIMIT 3
''')
top_3 = cursor.fetchall()
print("Top 3 sinh viên cao nhất:")
for row in top_3:
    print(row)

# Top 3 thấp nhất
cursor.execute('''
SELECT student_id, name, score
FROM student
ORDER BY score ASC
LIMIT 3
''')
bottom_3 = cursor.fetchall()
print("Top 3 sinh viên thấp nhất:")
for row in bottom_3:
    print(row)

Top 3 sinh viên cao nhất:
(2, 'Tran Thi Lan', 9.2)
(7, 'Bui Tien Dung', 9.2)
(3, 'Pham Van Nam', 7.9)
Top 3 sinh viên thấp nhất:
(1, 'Nguyen Minh Hoang', 6.7)
(10, 'Cao Thi Hanh', 7.0)
(9, 'Duong Huu Phuc', 7.2)


### **5.  Hãy bổ sung thêm một trường graduation_date có kiểu dữ liệu là DATETIME vào bảng student để xác định thời gian tốt nghiệp của từng bạn, trong đó thời gian tốt nghiệp được xác định bởi thời gian hiện tại cộng với số hạng tương ứng của bạn đó tính theo điểm số**

In [ ]:
# Bước 1: Thêm cột graduation_date vào bảng student 
cursor.execute('''
ALTER TABLE student ADD COLUMN graduation_date TEXT
''')
conn.commit()

# Bước 2: Tạo bảng tạm chứa thông tin xếp hạng
cursor.execute('''
CREATE TEMPORARY TABLE temp_rank AS
SELECT student_id,
       RANK() OVER (ORDER BY score DESC) AS rank
FROM student
''')
conn.commit()

# Bước 3: Cập nhật graduation_date dựa trên bảng tạm
cursor.execute('''
UPDATE student
SET graduation_date = DATE('now', '+' || (SELECT rank FROM temp_rank WHERE temp_rank.student_id = student.student_id) || ' days')
''')
conn.commit()



In [27]:
# Bước 4: Hiển thị kết quả
cursor.execute('SELECT * FROM student')
updated_students = cursor.fetchall()
print("Bảng student với graduation_date:")
for row in updated_students:
    print(row)

conn.close

Bảng student với graduation_date:
(1, 'Nguyen Minh Hoang', 'May Tinh', 12, 6.7, '2025-03-20')
(2, 'Tran Thi Lan', 'Kinh Te', 34, 9.2, '2025-03-15')
(3, 'Pham Van Nam', 'Toan Tin', 26, 7.9, '2025-03-17')
(7, 'Bui Tien Dung', 'Kinh Te', 34, 9.2, '2025-03-15')
(9, 'Duong Huu Phuc', 'Kinh Te', 26, 7.2, '2025-03-18')
(10, 'Cao Thi Hanh', 'May Tinh', 26, 7.0, '2025-03-19')


<function Connection.close()>